In [13]:
import numpy as np
import pandas as pd
from env import get_db_url

In [14]:
query = '''
SELECT *
FROM employees
LIMIT 200;

'''
employee_dept = pd.read_sql(query, get_db_url('employees'))
employee_dept.head()

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

5. a. Intentionally make a typo in the database url. What kind of error message do you see?

In [15]:
query = '''
SELECT *
FROM employees
LIMIT 200;

'''
employee_dept = pd.read_sql(query, get_db_url('employes')) #Traceback Error and access denied
employee_dept.head()

OperationalError: (pymysql.err.OperationalError) (1044, "Access denied for user 'quintela_2247'@'%' to database 'employes'")
(Background on this error at: https://sqlalche.me/e/14/e3q8)

b. Intentionally make an error in your SQL query. What does the error message look like?

In [16]:
query = '''
SEECT *
FROM employes
LIMIT 200;

'''
employee_dept = pd.read_sql(query, get_db_url('employees'))
employee_dept.head()

ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'SEECT *\nFROM employes\nLIMIT 200' at line 1")
[SQL: 
SEECT *
FROM employes
LIMIT 200;

]
(Background on this error at: https://sqlalche.me/e/14/f405)

6. Read the employees and titles tables into two separate DataFrames.

In [20]:
query = '''
SELECT *
FROM employees;

'''
employees_df = pd.read_sql(query, get_db_url('employees'))
employees_df.head()

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

In [21]:
query = '''
SELECT *
FROM titles;

'''
titles_df = pd.read_sql(query, get_db_url('employees'))
titles_df.head()

emp_no            title   from_date     to_date
0   10001  Senior Engineer  1986-06-26  9999-01-01
1   10002            Staff  1996-08-03  9999-01-01
2   10003  Senior Engineer  1995-12-03  9999-01-01
3   10004         Engineer  1986-12-01  1995-12-01
4   10004  Senior Engineer  1995-12-01  9999-01-01

7. How many rows and columns do you have in each DataFrame? Is that what you expected?

In [22]:
employees_df.shape #read (number_of_rows, number_of_columns)

(300024, 6)

In [23]:
titles_df.shape

(443308, 4)

8. Display the summary statistics for each DataFrame.

In [24]:
employees_df.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [25]:
titles_df.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

9. How many unique titles are in the titles DataFrame?

In [26]:
employees_df.nunique() #nope

emp_no        300024
birth_date      4750
first_name      1275
last_name       1637
gender             2
hire_date       5434
dtype: int64

In [27]:
titles_df['title'].nunique() #7 unique titles

7

10. What is the oldest date in the to_date column?

In [28]:
titles_df.sort_values(by='to_date').head() #1985-03-01

emp_no            title   from_date     to_date
16064    20869         Engineer  1985-02-17  1985-03-01
406358  474918         Engineer  1985-02-10  1985-03-11
346216  434232            Staff  1985-02-26  1985-03-20
35527    34059  Senior Engineer  1985-02-10  1985-03-23
277603  287732            Staff  1985-04-14  1985-04-17

In [29]:
titles_df['to_date'].min() #same

datetime.date(1985, 3, 1)

11. What is the most recent date in the to_date column?

In [30]:
titles_df['to_date'].max() #which means still working

datetime.date(9999, 1, 1)

#  Part # 2

1.  Copy the users and roles DataFrames from the examples above.

In [31]:
# Create the users DataFrame.

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users


id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [32]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles


id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

2.  What is the result of using a right join on the DataFrames?

In [36]:
users.merge(roles, left_on='role_id', right_on= 'id', how='left', indicator = True)  

id_x name_x  role_id  id_y    name_y     _merge
0     1    bob      1.0   1.0     admin       both
1     2    joe      2.0   2.0    author       both
2     3  sally      3.0   3.0  reviewer       both
3     4   adam      3.0   3.0  reviewer       both
4     5   jane      NaN   NaN       NaN  left_only
5     6   mike      NaN   NaN       NaN  left_only

It returns the appropriate information from roles discarding what is not needed

In [37]:
users.merge(roles, left_on='role_id', right_on= 'id', how='right', indicator = True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0     1      admin        both
1   2.0    joe      2.0     2     author        both
2   3.0  sally      3.0     3   reviewer        both
3   4.0   adam      3.0     3   reviewer        both
4   NaN    NaN      NaN     4  commenter  right_only

It maintains the fidelity of the roles table and adds appropriate info from users.  It discards the rest of the lines from users.

3.  What is the result of using an outer join on the DataFrames?

In [38]:
users.merge(roles, left_on='role_id', right_on= 'id', how='outer', indicator = True) 

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only

It gives us the 7 possibilities, the first six from users and the left over 7th possibility from the commenter that is unmatched.  Since admin author and reviewer appears on the users table it does not return those from roles.

4.  What happens if you drop the foreign keys from the DataFrames and try to merge them?

In [39]:
users.merge(roles, how='outer', indicator = True) 

id       name  role_id      _merge
0   1        bob      1.0   left_only
1   2        joe      2.0   left_only
2   3      sally      3.0   left_only
3   4       adam      3.0   left_only
4   5       jane      NaN   left_only
5   6       mike      NaN   left_only
6   1      admin      NaN  right_only
7   2     author      NaN  right_only
8   3   reviewer      NaN  right_only
9   4  commenter      NaN  right_only

It gives us the 9 total possibilities by combining the two tables.  since we are not joining on id and role_id respectively it gives us all the possibilities

5.  Load the mpg dataset from PyDataset.

In [40]:
from pydataset import data
mpg_df = data('mpg')
mpg_df

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

6.  Output and read the documentation for the mpg dataset.

In [41]:
data('mpg', show_doc=True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




7.  How many rows and columns are in the dataset?

There are 234 rows and 11 columns.  alternatively I can use .shape

In [141]:
mpg_df.shape

(234, 11)

8.  Check out your column names and perform any cleanup you may want on them.

In [42]:
mpg_df = mpg_df.rename(columns = {'displ': 'cylinders',
                        'cyl': 'cylinders',
                        'cty': 'city',
                        'hwy': 'highway',
                        'drv': 'drive',
                        'trans': 'transmission',
                        })
mpg_df

manufacturer   model  cylinders  year  cylinders transmission drive  city  \
1           audi      a4        1.8  1999          4     auto(l5)     f    18   
2           audi      a4        1.8  1999          4   manual(m5)     f    21   
3           audi      a4        2.0  2008          4   manual(m6)     f    20   
4           audi      a4        2.0  2008          4     auto(av)     f    21   
5           audi      a4        2.8  1999          6     auto(l5)     f    16   
..           ...     ...        ...   ...        ...          ...   ...   ...   
230   volkswagen  passat        2.0  2008          4     auto(s6)     f    19   
231   volkswagen  passat        2.0  2008          4   manual(m6)     f    21   
232   volkswagen  passat        2.8  1999          6     auto(l5)     f    16   
233   volkswagen  passat        2.8  1999          6   manual(m5)     f    18   
234   volkswagen  passat        3.6  2008          6     auto(s6)     f    17   

     highway fl    class  
1         29  p  compact  
2         29  p  compact  
3         31  p  compact  
4         30  p  compact  
5         26  p  compact  
..       ... ..      ...  
230       28  p  midsize  
231       29  p  midsize  
232       26  p  midsize  
233       26  p  midsize  
234       26  p  midsize  

[234 rows x 11 columns]

9.  Display the summary statistics for the dataset.

In [43]:
mpg_df.describe()

cylinders         year   cylinders        city     highway
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

10.  How many different manufacturers are there?

In [204]:
mpg_df.value_counts('manufacturer').count()

15

11.  How many different models are there?

In [45]:
mpg_df.value_counts('model').count()

38

12.  Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

In [46]:
mpg_df = mpg_df.assign(mileage_difference = mpg_df.highway - mpg_df.city)
mpg_df

manufacturer   model  cylinders  year  cylinders transmission drive  city  \
1           audi      a4        1.8  1999          4     auto(l5)     f    18   
2           audi      a4        1.8  1999          4   manual(m5)     f    21   
3           audi      a4        2.0  2008          4   manual(m6)     f    20   
4           audi      a4        2.0  2008          4     auto(av)     f    21   
5           audi      a4        2.8  1999          6     auto(l5)     f    16   
..           ...     ...        ...   ...        ...          ...   ...   ...   
230   volkswagen  passat        2.0  2008          4     auto(s6)     f    19   
231   volkswagen  passat        2.0  2008          4   manual(m6)     f    21   
232   volkswagen  passat        2.8  1999          6     auto(l5)     f    16   
233   volkswagen  passat        2.8  1999          6   manual(m5)     f    18   
234   volkswagen  passat        3.6  2008          6     auto(s6)     f    17   

     highway fl    class  mileage_difference  
1         29  p  compact                  11  
2         29  p  compact                   8  
3         31  p  compact                  11  
4         30  p  compact                   9  
5         26  p  compact                  10  
..       ... ..      ...                 ...  
230       28  p  midsize                   9  
231       29  p  midsize                   8  
232       26  p  midsize                  10  
233       26  p  midsize                   8  
234       26  p  midsize                   9  

[234 rows x 12 columns]

13.  Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

In [47]:
mpg_df = mpg_df.assign(average_mileage = (mpg_df.city + mpg_df.highway)/2)
mpg_df

manufacturer   model  cylinders  year  cylinders transmission drive  city  \
1           audi      a4        1.8  1999          4     auto(l5)     f    18   
2           audi      a4        1.8  1999          4   manual(m5)     f    21   
3           audi      a4        2.0  2008          4   manual(m6)     f    20   
4           audi      a4        2.0  2008          4     auto(av)     f    21   
5           audi      a4        2.8  1999          6     auto(l5)     f    16   
..           ...     ...        ...   ...        ...          ...   ...   ...   
230   volkswagen  passat        2.0  2008          4     auto(s6)     f    19   
231   volkswagen  passat        2.0  2008          4   manual(m6)     f    21   
232   volkswagen  passat        2.8  1999          6     auto(l5)     f    16   
233   volkswagen  passat        2.8  1999          6   manual(m5)     f    18   
234   volkswagen  passat        3.6  2008          6     auto(s6)     f    17   

     highway fl    class  mileage_difference  average_mileage  
1         29  p  compact                  11             23.5  
2         29  p  compact                   8             25.0  
3         31  p  compact                  11             25.5  
4         30  p  compact                   9             25.5  
5         26  p  compact                  10             21.0  
..       ... ..      ...                 ...              ...  
230       28  p  midsize                   9             23.5  
231       29  p  midsize                   8             25.0  
232       26  p  midsize                  10             21.0  
233       26  p  midsize                   8             22.0  
234       26  p  midsize                   9             21.5  

[234 rows x 13 columns]

14.  Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.

In [48]:
mpg_df = mpg_df.assign(is_automatic = mpg_df.transmission.str.contains('auto'))
mpg_df

manufacturer   model  cylinders  year  cylinders transmission drive  city  \
1           audi      a4        1.8  1999          4     auto(l5)     f    18   
2           audi      a4        1.8  1999          4   manual(m5)     f    21   
3           audi      a4        2.0  2008          4   manual(m6)     f    20   
4           audi      a4        2.0  2008          4     auto(av)     f    21   
5           audi      a4        2.8  1999          6     auto(l5)     f    16   
..           ...     ...        ...   ...        ...          ...   ...   ...   
230   volkswagen  passat        2.0  2008          4     auto(s6)     f    19   
231   volkswagen  passat        2.0  2008          4   manual(m6)     f    21   
232   volkswagen  passat        2.8  1999          6     auto(l5)     f    16   
233   volkswagen  passat        2.8  1999          6   manual(m5)     f    18   
234   volkswagen  passat        3.6  2008          6     auto(s6)     f    17   

     highway fl    class  mileage_difference  average_mileage  is_automatic  
1         29  p  compact                  11             23.5          True  
2         29  p  compact                   8             25.0         False  
3         31  p  compact                  11             25.5         False  
4         30  p  compact                   9             25.5          True  
5         26  p  compact                  10             21.0          True  
..       ... ..      ...                 ...              ...           ...  
230       28  p  midsize                   9             23.5          True  
231       29  p  midsize                   8             25.0         False  
232       26  p  midsize                  10             21.0          True  
233       26  p  midsize                   8             22.0         False  
234       26  p  midsize                   9             21.5          True  

[234 rows x 14 columns]

15.  Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average?

In [49]:
mpg_df.groupby(['manufacturer']).average_mileage.describe().sort_values(by='mean', ascending=False).head(1)  #try to figure out how to isolate the max mean


count  mean       std   min   25%   50%   75%   max
manufacturer                                                     
honda           9.0  28.5  1.984313  25.0  28.0  28.5  30.0  30.5

16.  Do automatic or manual cars have better miles per gallon?

In [50]:
mpg_df.groupby('is_automatic').average_mileage.describe().sort_values(by='mean', ascending=False).head(1)

count       mean       std   min   25%   50%   75%   max
is_automatic                                                          
False          77.0  22.227273  5.160927  10.5  19.5  22.0  25.0  39.5

Generally manual transmissions have better gas mileage by quartiles max mean etc.  it seems at best in the min there may be some overlap.  Further analysis is needed.

# Part 3

1.  Use your get_db_url function to help you explore the data from the chipotle database.

In [66]:
chipotle_df = pd.read_sql('SELECT * FROM orders;', get_db_url('chipotle'))
chipotle_df

id  order_id  quantity                              item_name  \
0        1         1         1           Chips and Fresh Tomato Salsa   
1        2         1         1                                   Izze   
2        3         1         1                       Nantucket Nectar   
3        4         1         1  Chips and Tomatillo-Green Chili Salsa   
4        5         2         2                           Chicken Bowl   
...    ...       ...       ...                                    ...   
4617  4618      1833         1                          Steak Burrito   
4618  4619      1833         1                          Steak Burrito   
4619  4620      1834         1                     Chicken Salad Bowl   
4620  4621      1834         1                     Chicken Salad Bowl   
4621  4622      1834         1                     Chicken Salad Bowl   

                                                                     choice_description  \
0                                                                                   nan   
1                                                                          [Clementine]   
2                                                                               [Apple]   
3                                                                                   nan   
4            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4618               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4620                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4621                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

     item_price  
0        $2.39   
1        $3.39   
2        $3.39   
3        $2.39   
4       $16.98   
...         ...  
4617    $11.75   
4618    $11.75   
4619    $11.25   
4620     $8.75   
4621     $8.75   

[4622 rows x 6 columns]

In [67]:
data('chipotle', show_doc=True)

Did you mean:
Chile, Chirot, Cowles, hirose, cities, Hitters, shuttle, pottery, Pottery, alcohol, coleman


2.  What is the total price for each order?

In [69]:
chipotle_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   id                  4622 non-null   int64 
 1   order_id            4622 non-null   int64 
 2   quantity            4622 non-null   int64 
 3   item_name           4622 non-null   object
 4   choice_description  4622 non-null   object
 5   item_price          4622 non-null   object
dtypes: int64(3), object(3)
memory usage: 216.8+ KB


In [70]:
chipotle_df.item_price.str.strip('$').astype('float')

0        2.39
1        3.39
2        3.39
3        2.39
4       16.98
        ...  
4617    11.75
4618    11.75
4619    11.25
4620     8.75
4621     8.75
Name: item_price, Length: 4622, dtype: float64

In [71]:
total_price_df = chipotle_df.assign(total_price = (chipotle_df.quantity * chipotle_df.item_price.str.strip('$').astype('float')))

In [76]:
total_price_df #in new appended column

id  order_id  quantity                              item_name  \
0        1         1         1           Chips and Fresh Tomato Salsa   
1        2         1         1                                   Izze   
2        3         1         1                       Nantucket Nectar   
3        4         1         1  Chips and Tomatillo-Green Chili Salsa   
4        5         2         2                           Chicken Bowl   
...    ...       ...       ...                                    ...   
4617  4618      1833         1                          Steak Burrito   
4618  4619      1833         1                          Steak Burrito   
4619  4620      1834         1                     Chicken Salad Bowl   
4620  4621      1834         1                     Chicken Salad Bowl   
4621  4622      1834         1                     Chicken Salad Bowl   

                                                                     choice_description  \
0                                                                                   nan   
1                                                                          [Clementine]   
2                                                                               [Apple]   
3                                                                                   nan   
4            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4618               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4620                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4621                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

     item_price  total_price  
0        $2.39          2.39  
1        $3.39          3.39  
2        $3.39          3.39  
3        $2.39          2.39  
4       $16.98         33.96  
...         ...          ...  
4617    $11.75         11.75  
4618    $11.75         11.75  
4619    $11.25         11.25  
4620     $8.75          8.75  
4621     $8.75          8.75  

[4622 rows x 7 columns]

3.  What are the most popular 3 items?

In [105]:
total_price_df.item_name.value_counts().head(3) 

Chicken Bowl           726
Chicken Burrito        553
Chips and Guacamole    479
Name: item_name, dtype: int64

In [224]:
chipotle_df.groupby('item_name').quantity.sum().sort_values(ascending=False).head(3)

item_name
Chicken Bowl           761
Chicken Burrito        591
Chips and Guacamole    506
Name: quantity, dtype: int64

4.  Which item has produced the most revenue?

In [134]:
total_pivot = total_price_df.pivot_table(index='item_name', columns='quantity', values = 'total_price', aggfunc = 'sum', margins = True)
total_pivot

quantity                                      1        2        3       4  \
item_name                                                                   
6 Pack Soft Drink                        343.97    25.96      NaN     NaN   
Barbacoa Bowl                            672.36      NaN      NaN     NaN   
Barbacoa Burrito                         894.75      NaN      NaN     NaN   
Barbacoa Crispy Tacos                    101.71    37.00      NaN     NaN   
Barbacoa Salad Bowl                      106.40      NaN      NaN     NaN   
Barbacoa Soft Tacos                      250.46      NaN      NaN     NaN   
Bottled Water                            199.48    67.08    73.62   48.00   
Bowl                                       7.40      NaN    66.60     NaN   
Burrito                                   44.40      NaN      NaN     NaN   
Canned Soda                               91.56    82.84      NaN   17.44   
Canned Soft Drink                        330.00   135.00    78.75   60.00   
Carnitas Bowl                            677.96    47.00   105.75     NaN   
Carnitas Burrito                         579.33    37.00      NaN     NaN   
Carnitas Crispy Tacos                     59.98    35.96      NaN     NaN   
Carnitas Salad                             8.99      NaN      NaN     NaN   
Carnitas Salad Bowl                       66.34      NaN      NaN     NaN   
Carnitas Soft Tacos                      375.94      NaN      NaN     NaN   
Chicken Bowl                            6700.02  1167.04   177.57     NaN   
Chicken Burrito                         4957.08   992.48   157.50  280.00   
Chicken Crispy Tacos                     420.15   103.96      NaN     NaN   
Chicken Salad                             81.09      NaN      NaN     NaN   
Chicken Salad Bowl                       951.25   555.00      NaN     NaN   
Chicken Soft Tacos                      1017.17   181.84      NaN     NaN   
Chips                                    414.79   146.20    19.35     NaN   
Chips and Fresh Tomato Salsa             287.61    35.40      NaN   47.20   
Chips and Guacamole                     1975.41   388.96    40.05   71.20   
Chips and Mild Fresh Tomato Salsa          3.00      NaN      NaN     NaN   
Chips and Roasted Chili Corn Salsa        61.95    11.80      NaN     NaN   
Chips and Roasted Chili-Corn Salsa        43.02      NaN      NaN     NaN   
Chips and Tomatillo Green Chili Salsa    120.95    23.60      NaN     NaN   
Chips and Tomatillo Red Chili Salsa      135.70    23.60      NaN     NaN   
Chips and Tomatillo-Green Chili Salsa     69.31    19.12      NaN     NaN   
Chips and Tomatillo-Red Chili Salsa       40.63     9.56    43.02     NaN   
Crispy Tacos                              14.80      NaN      NaN     NaN   
Izze                                      67.80      NaN      NaN     NaN   
Nantucket Nectar                          84.75    27.12      NaN     NaN   
Salad                                     14.80      NaN      NaN     NaN   
Side of Chips                            167.31      NaN    15.21     NaN   
Steak Bowl                              2066.64   334.96    78.21     NaN   
Steak Burrito                           3522.23   547.40   166.50     NaN   
Steak Crispy Tacos                       339.36    35.96      NaN     NaN   
Steak Salad                               35.66      NaN      NaN     NaN   
Steak Salad Bowl                         296.03    95.12      NaN     NaN   
Steak Soft Tacos                         517.55    37.00      NaN     NaN   
Veggie Bowl                              834.03    67.92      NaN     NaN   
Veggie Burrito                           901.02      NaN   101.25     NaN   
Veggie Crispy Tacos                        8.49      NaN      NaN     NaN   
Veggie Salad                              50.94      NaN      NaN     NaN   
Veggie Salad Bowl                        182.50      NaN      NaN     NaN   
Veggie Soft Tacos                         56.98    33.96      NaN     NaN

In [183]:
total_pivot[total_pivot.All > 1000]#.sort_values(by='All')

quantity                             1        2        3       4     5     7  \
item_name                                                                      
Chicken Bowl                   6700.02  1167.04   177.57     NaN   NaN   NaN   
Chicken Burrito                4957.08   992.48   157.50  280.00   NaN   NaN   
Chicken Salad Bowl              951.25   555.00      NaN     NaN   NaN   NaN   
Chicken Soft Tacos             1017.17   181.84      NaN     NaN   NaN   NaN   
Chips and Fresh Tomato Salsa    287.61    35.40      NaN   47.20   NaN   NaN   
Chips and Guacamole            1975.41   388.96    40.05   71.20   NaN   NaN   
Steak Bowl                     2066.64   334.96    78.21     NaN   NaN   NaN   
Steak Burrito                  3522.23   547.40   166.50     NaN   NaN   NaN   
Veggie Burrito                  901.02      NaN   101.25     NaN   NaN   NaN   
All                           31251.05  5305.84  1123.38  523.84  37.5  73.5   

quantity                           8     10      15       All  
item_name                                                      
Chicken Bowl                     NaN    NaN     NaN   8044.63  
Chicken Burrito                  NaN    NaN     NaN   6387.06  
Chicken Salad Bowl               NaN    NaN     NaN   1506.25  
Chicken Soft Tacos               NaN    NaN     NaN   1199.01  
Chips and Fresh Tomato Salsa     NaN    NaN  663.75   1033.96  
Chips and Guacamole              NaN    NaN     NaN   2475.62  
Steak Bowl                       NaN    NaN     NaN   2479.81  
Steak Burrito                    NaN    NaN     NaN   4236.13  
Veggie Burrito                   NaN    NaN     NaN   1002.27  
All                           108.16  150.0  663.75  39237.02

In [197]:
total_pivot.All.values[:-1].max()

8044.63

In [112]:
pd.crosstab(total_price_df.item_name, total_price_df.total_price, margins=True) #Deadend but think about it later

total_price                            1.09  1.25  1.5  1.69  1.99  2.15  \
item_name                                                                  
6 Pack Soft Drink                         0     0    0     0     0     0   
Barbacoa Bowl                             0     0    0     0     0     0   
Barbacoa Burrito                          0     0    0     0     0     0   
Barbacoa Crispy Tacos                     0     0    0     0     0     0   
Barbacoa Salad Bowl                       0     0    0     0     0     0   
Barbacoa Soft Tacos                       0     0    0     0     0     0   
Bottled Water                            22     0  117     0     0     0   
Bowl                                      0     0    0     0     0     0   
Burrito                                   0     0    0     0     0     0   
Canned Soda                              84     0    0     0     0     0   
Canned Soft Drink                         0   264    0     0     0     0   
Carnitas Bowl                             0     0    0     0     0     0   
Carnitas Burrito                          0     0    0     0     0     0   
Carnitas Crispy Tacos                     0     0    0     0     0     0   
Carnitas Salad                            0     0    0     0     0     0   
Carnitas Salad Bowl                       0     0    0     0     0     0   
Carnitas Soft Tacos                       0     0    0     0     0     0   
Chicken Bowl                              0     0    0     0     0     0   
Chicken Burrito                           0     0    0     0     0     0   
Chicken Crispy Tacos                      0     0    0     0     0     0   
Chicken Salad                             0     0    0     0     0     0   
Chicken Salad Bowl                        0     0    0     0     0     0   
Chicken Soft Tacos                        0     0    0     0     0     0   
Chips                                     0     0    0     0     1   192   
Chips and Fresh Tomato Salsa              0     0    0     0     0     0   
Chips and Guacamole                       0     0    0     0     0     0   
Chips and Mild Fresh Tomato Salsa         0     0    0     0     0     0   
Chips and Roasted Chili Corn Salsa        0     0    0     0     0     0   
Chips and Roasted Chili-Corn Salsa        0     0    0     0     0     0   
Chips and Tomatillo Green Chili Salsa     0     0    0     0     0     0   
Chips and Tomatillo Red Chili Salsa       0     0    0     0     0     0   
Chips and Tomatillo-Green Chili Salsa     0     0    0     0     0     0   
Chips and Tomatillo-Red Chili Salsa       0     0    0     0     0     0   
Crispy Tacos                              0     0    0     0     0     0   
Izze                                      0     0    0     0     0     0   
Nantucket Nectar                          0     0    0     0     0     0   
Salad                                     0     0    0     0     0     0   
Side of Chips                             0     0    0    99     0     0   
Steak Bowl                                0     0    0     0     0     0   
Steak Burrito                             0     0    0     0     0     0   
Steak Crispy Tacos                        0     0    0     0     0     0   
Steak Salad                               0     0    0     0     0     0   
Steak Salad Bowl                          0     0    0     0     0     0   
Steak Soft Tacos                          0     0    0     0     0     0   
Veggie Bowl                               0     0    0     0     0     0   
Veggie Burrito                            0     0    0     0     0     0   
Veggie Crispy Tacos                       0     0    0     0     0     0   
Veggie Salad                              0     0    0     0     0     0   
Veggie Salad Bowl                         0     0    0     0     0     0   
Veggie Soft Tacos                         0     0    0     0     0     0   
All                                     106   26

In [109]:
total_price_df.item_name.value_counts

<bound method IndexOpsMixin.value_counts of 0                Chips and Fresh Tomato Salsa
1                                        Izze
2                            Nantucket Nectar
3       Chips and Tomatillo-Green Chili Salsa
4                                Chicken Bowl
                        ...                  
4617                            Steak Burrito
4618                            Steak Burrito
4619                       Chicken Salad Bowl
4620                       Chicken Salad Bowl
4621                       Chicken Salad Bowl
Name: item_name, Length: 4622, dtype: object>

5.  Join the employees and titles DataFrames together.

In [201]:
#left_df.merge(right_df, how='inner', on=None, left_on=None, right_on=None, left_index=False, right_index=False, indicator=False)
merged_table = employees_df.merge(titles_df, left_on='emp_no', right_on='emp_no', how='outer')
merged_table

emp_no  birth_date first_name last_name gender   hire_date  \
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26   
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21   
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28   
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
4        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
...        ...         ...        ...       ...    ...         ...   
443303  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443304  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443305  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443306  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443307  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30   

                  title   from_date     to_date  
0       Senior Engineer  1986-06-26  9999-01-01  
1                 Staff  1996-08-03  9999-01-01  
2       Senior Engineer  1995-12-03  9999-01-01  
3              Engineer  1986-12-01  1995-12-01  
4       Senior Engineer  1995-12-01  9999-01-01  
...                 ...         ...         ...  
443303         Engineer  1987-08-30  1992-08-29  
443304  Senior Engineer  1992-08-29  9999-01-01  
443305     Senior Staff  1998-12-27  9999-01-01  
443306            Staff  1993-12-27  1998-12-27  
443307         Engineer  1997-11-30  9999-01-01  

[443308 rows x 9 columns]

6.  For each title, find the hire date of the employee that was hired most recently with that title.

In [231]:
merged_table.groupby('title').hire_date.agg('max')

title
Assistant Engineer    1999-12-24
Engineer              2000-01-28
Manager               1992-02-05
Senior Engineer       2000-01-01
Senior Staff          2000-01-13
Staff                 2000-01-12
Technique Leader      1999-12-31
Name: hire_date, dtype: object

7.  Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)

In [230]:
pd.crosstab(merged_table.title, merged_table.department)

AttributeError: 'DataFrame' object has no attribute 'department'